### ISSUE TITLE
2.1.1.0-1 dict parsing errors

### ISSUE THREAD

**Issue by n2aws** [ansible/ansible#17001]
Edited,  didn't realize I had half of this report listed within comment blocks,  so they didn't appear.  Apologies!
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

```
ansible 2.1.1.0
  config file = /home/jpalmer/source/s2s/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```
##### OS / ENVIRONMENT

Ubuntu Xenial
##### SUMMARY

I have a group_vars/all/foo dict that has not caused any issues until 2.1.1.0-1.  With this version, the dict seems to be incorrectly parsed, even if the playbook itself isn't utilizing the dict.  I get several errors about missing files (users authorized_keys pubfiles, that are contained as lookups in the dict)
##### STEPS TO REPRODUCE

1) Have a dict in group_vars/all that contains a lookup to a file (in my case, an ssh authorized_keys file) 
2) Make sure one of the files doesn't exist.
3) Have a playbook that doesn't reference the dict
4) Execute the playbook, and look for errors about missing files, even when your playbook doesn't reference them. 

##### EXPECTED RESULTS

dict parsing to continue working the way it used to.  ie, evaluating a lookup only as needed,  rather than lookup all files for every run regardless of if they are needed or not.
##### ACTUAL RESULTS

The dict is parsed fully for every run, regardless if the playbook needs it.

**Comment by alikins** [comment]
Have an example playbook that demonstrates this?

**Comment by n2aws** [comment]
I do,   but it doesn't appear to be playbook specific.

group_vars/all/user,     I have a dict in the form of:

```
---
users:
  ops:
    foouser:
      name: Foo User 
      groups: ops
      ssh_pubkey: "{{ lookup('file', 'foouser.pub') }}"
```

<snipped for brevity, but I have about 4 groups (like ops)  and about 40 users in total in my dict.>

playbook:

```
---
- hosts: [mysql] 
  remote_user: myuser 
  gather_facts: false
  become: True
  tasks:
    - name: Add ops group
      group: name=ops state=present

    - name: Add newdba to machine 
      user: name=newdba groups=ops,mysql append=yes

    - name: ssh key
      authorized_key: user=newdba key="ssh-rsa AAAAB3NzaC1kc3MAAACBAMvS0oHc8O6tBn<...>"
```

And finally,   in the hostgroup:

```
[mysql]
server1-atl
192.168.0.37
server4-chi
192.168.15.38
```

**Comment by n2aws** [comment]
Not that context is necessary,   the normal dict is used by a role we've created to manage users on various machines.   This particular playbook was a 1off to add a new DBA to all of our mysql servers at once, without performing a full ansible run against the fleet.

This same setup worked prior to upgrading from a previous 2.1.x version,  to 2.1.1.0-1 this morning.

**Comment by agaffney** [comment]
To add to this, there was someone in #ansible a few days ago showing what appeared to be a non-lazy evaluation of lookup() inside group_vars that weren't referenced by anything in the play. It exhibited as ansible taking much longer to startup than it had in previous versions. It also started with the upgrade to 2.1.1.0

A general problem with vars being evaluated at parse time rather than when referenced could probably account for both of these issues.

**Comment by dagwieers** [comment]
Hmm, learned something new today. Didn't know that **hosts:** can be a list.

**Comment by n2aws** [comment]
@jimi-c @alikins  I see this is still flagged with "needs_info"  however,  I have provided the info that was asked for.   Is there still info needed?

**Comment by jimi-c** [comment]
@palmerit sorry no more info needed, we just hadn't followed up with updating the labels.

Thanks!

**Comment by jimi-c** [comment]
@palmerit sorry for the late follow-up, but this was corrected in 2.1.2.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
